pkg/kapp/app/recorded_app.go

```diff
@@ -196,27 +191,83 @@ func (a *RecordedApp) createOrUpdate(c *corev1.ConfigMap, labels map[string]stri
 				return fmt.Errorf("Getting app: %s", err)
 			}
 
-			err = a.mergeAppUpdates(existingConfigMap, labels)
-			if err != nil {
-				return err
-			}
+			return a.updateApp(existingConfigMap, labels)
+		}
 
-			_, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Update(context.TODO(), existingConfigMap, metav1.UpdateOptions{})
-			if err != nil {
-				return fmt.Errorf("Updating app: %s", err)
+		return fmt.Errorf("Creating app: %s", err)
+	}
+
+	return nil
+}
+
+func (a *RecordedApp) updateApp(existingConfigMap *corev1.ConfigMap, labels map[string]string) error {
+	err := a.mergeAppUpdates(existingConfigMap, labels)
+	if err != nil {
+		return err
+	}
+
+	_, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Update(context.TODO(), existingConfigMap, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("Updating app: %s", err)
+	}
+
+	return nil
+}
+
+func (a *RecordedApp) RenamePrevApp(prevAppName string, labels map[string]string) error {
+	defer a.logger.DebugFunc("RenamePrevApp").Finish()
+	var c *corev1.ConfigMap
+	var err error
+
+	if a.isMigrationEnabled() {
+		c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.fqName, metav1.GetOptions{})
+		if err == nil {
+			return a.updateApp(c, labels)
+		} else if err != nil {
+			if errors.IsNotFound(err) {
+				c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.name, metav1.GetOptions{})
+				if err == nil {
+					return a.migrate(c, labels, a.fqName)
+				}
 			}
+		}
+	} else {
+		c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.name, metav1.GetOptions{})
+		if err == nil {
+			return a.updateApp(c, labels)
+		}
+	}
 
-			return nil
+	if errors.IsNotFound(err) {
+		if a.isMigrationEnabled() {
+			c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName+AppSuffix, metav1.GetOptions{})
+			if err == nil {
+				return a.migrate(c, labels, a.fqName)
+			} else if err != nil {
+				if errors.IsNotFound(err) {
+					c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName, metav1.GetOptions{})
+					if err == nil {
+						return a.migrate(c, labels, a.fqName)
+					}
+				}
+			}
+		} else {
+			c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName, metav1.GetOptions{})
+			if err == nil {
+				return a.migrate(c, labels, a.name)
```

Comment:
I think this would add the `KappIsConfigmapMigratedAnnotationKey` to the configmap even though migration is disabled. Maybe it was best to keep the `mergeAppAnnotationUpdates()` outside the migrate function? 
(Although I did like that right now we don't have to call it every time we use `migrate()`)

Reply:
This is a great callout, I updated the tests to all check for the presence / absence of that annotation and I updated this path to not use the `migrate()` function - thus keeping the annotation addition in that function.